### PR TITLE
Add libcudf strings split API that accepts regex pattern

### LINKED ISSUE #3584: [FEA] Support regular expression patterns for string split
I'd like to be able to pass a regular expression pattern for string splitting, like I could do in base Python or pandas. This is relevant for tasks like trying to do sentence tokenization with regular expressions, as sentences can end with more than one type of punctuation (for example).

```python
import pandas
import cudf
import re
​
print(s)
0    this is. an? example
dtype: object

s = cudf.Series(['this is. an? example'])
ps = s.to_pandas()
​
print(ps.str.split('[.?!] ', expand=True))
         0   1        2
0  this is  an  example

print(re.split('[.?!] ', ps.iloc[0]))
['this is', 'an', 'example']

# cudf
print(s.str.split('[.?!] '))
                      0
0  this is. an? example

```

### PR DESCRIPTION
**Pull request by davidwendt**
Reference #3584

This PR adds 4 new libcudf strings APIs for split.
- `cudf::strings::split_re` - split using regex to locate delimiters with table output like `cudf::strings::split`.
- `cudf::strings::rsplit_re` - same as `split_re` but delimiter search starts from the end of each string
- `cudf::strings::split_record_re` - same as `split_re` but returns a list column like `split_record` does
- `cudf::strings::rsplit_record_re` - same as `split_record_re` but delimiter search starts from the end of each string

Like `split/rsplit` the results try to match Pandas behavior for these. The `record` results are similar to specifying `expand=False` in the Pandas `split/rsplit` APIs. Python/Cython updates for cuDF will be in a follow-on PR.
Currently, Pandas does not support regex for its `rsplit` even though it has been documented and there is an issue [here](https://github.com/pandas-dev/pandas/issues/29633).

New gtests have been added for these along with some additional tests that were missing for the non-regex versions of these APIs.

### CODE REVIEW

**Review by robertmaynard — APPROVED** [review]
CMake changes LGTM

**Review by ajschmidt8 — APPROVED** [review]
Approving `ops-codeowner` file changes

File: cpp/tests/strings/split_tests.cpp
Comment (ttnghia) [inline cpp/tests/strings/split_tests.cpp:357]: Can you add a test with `"\\."` pattern please? We need to verify the result for that case in Spark plugin side. Thanks.

Comment (davidwendt) [reply cpp/tests/strings/split_tests.cpp:357]: Can you provide some sample strings that you want to split?
Also, if you are trying to split on a literal '.' character, you should use the regular `split` and not the regex ones.

Comment (ttnghia) [reply cpp/tests/strings/split_tests.cpp:357]: Please check out the comment here: https://github.com/NVIDIA/spark-rapids/pull/4681/#discussion_r797928342

Comment (davidwendt) [reply cpp/tests/strings/split_tests.cpp:357]: Sorry, I'm not finding any example strings in the link provided. Maybe I'm just unable to see them.
You are welcome to pull this PR to try it out with the Spark testcases. 
I don't think we need a `\\.` testcase here since it would not be recommended to use regex for a single character.

Comment (ttnghia) [reply cpp/tests/strings/split_tests.cpp:357]: I copy the example from there to here:

> For example if I did a split with "\\." as the separator on "a.b.c" I would expect an output of ["a", "b", "c"]. But I cannot take "\\." and pass it into a non-regexp string split, because it will try to look for a "\\" character in front of the '.'. Am I wrong/missing something? I think we could still support it, but we would have to transpile the regular expression into a constant string.

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (ttnghia) [inline cpp/include/cudf/strings/split/split_re.hpp:74]: Wait, I just realize that here we have `std::string` pattern instead of `string_scalar` as the delimiter in `strings::split`. Is there a reason for that?

Comment (ttnghia) [reply cpp/include/cudf/strings/split/split_re.hpp:74]: If we have to use `std::string` then the interface of this new API is much different from `strings::split` and we can't use the same Java binding to switch between the two.

Comment (davidwendt) [reply cpp/include/cudf/strings/split/split_re.hpp:74]: The `pattern` is never copied to the device.
Perhaps you use the `string_scalar` ctor that takes `std::string` 
https://docs.rapids.ai/api/libcudf/stable/classcudf_1_1string__scalar.html#ab8531ebc14e2a9b5a8695b9a7e937c2d

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (bdice) [inline cpp/include/cudf/strings/split/split_re.hpp:215]:
```suggestion
 * @throw cudf::logic_error if `pattern` is empty.
```

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [inline cpp/src/strings/split/split_re.cu:94]: Does this line do anything? It looks like `end` is initialized with the value `d_str.length();`, so it should be fine to skip this unless `end` is modified by some other function that I can't identify.
```suggestion
```

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [inline cpp/src/strings/split/split_re.cu:101]:
```suggestion
      // update first entry -- this happens when maxsplit is hit before the end of the string
```

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [inline cpp/src/strings/split/split_re.cu:120]: Missing param:
```suggestion
 * @param direction Whether tokens are generated forwards or backwards.
 * @param max_tokens The maximum number of tokens for each split.
```

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [inline cpp/src/strings/split/split_re.cu:185]: I think `>=` is slightly clearer than `> ... - 1` in this case. Also I'd move the parentheses a little bit. The clarity of operator precedence might even warrant two sets of parentheses, if you still think the first set is helpful. I knew `>=` precedes `||` but did not know where `?` fell in the order of operations until I [looked it up](https://en.cppreference.com/w/cpp/language/operator_precedence).
```suggestion
    return (column_index >= token_count - 1 || d_strings.is_null(idx))
```

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [inline cpp/src/strings/split/split_re.cu:200]: This line could be moved into the `generate_tokens` function since it is repeated by both calling functions and is needed for correctness.

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [inline cpp/src/strings/split/split_re.cu:260]: This function's implementation is so nice to read, because all the complicated bits are abstracted into helper functions and everything has clear names and comments. Just wanted to say this made me happy to review!

**Comment by ttnghia** [comment]
Just add DO_NOT_MERGE label as we are testing it and potentially uncover some issues.

File: cpp/tests/strings/split_tests.cpp
Comment (bdice) [inline cpp/tests/strings/split_tests.cpp:308]: Tests look great overall. Should we consider tests involving multi-byte UTF-8 characters? Are those supported?

**Review by bdice — COMMENTED** [review]
Very nice overall. Comments attached.

File: cpp/src/strings/split/split_re.cu
Comment (davidwendt) [reply cpp/src/strings/split/split_re.cu:94]: The `begin` and `end` are currently modified by `prog.find()` -- that is why they are not `const` :-)
This behavior is on the list to fix in a later PR.

File: cpp/src/strings/split/split_re.cu
Comment (davidwendt) [reply cpp/src/strings/split/split_re.cu:101]: The `max-tokens` is actually more correct here. I feel it is more clear what is actually happening.

File: cpp/tests/strings/split_tests.cpp
Comment (davidwendt) [reply cpp/tests/strings/split_tests.cpp:308]: The `é` is a multi-byte UTF-8 character. I use it in the data and in the regex pattern.

File: cpp/src/strings/split/split_re.cu
Comment (bdice) [reply cpp/src/strings/split/split_re.cu:101]: Sounds good. I was 50/50 on this suggestion, just thought it was worth proposing.

File: cpp/tests/strings/split_tests.cpp
Comment (bdice) [reply cpp/tests/strings/split_tests.cpp:308]: Thanks for the correction. I had forgotten that é was multi-byte. I always think of more obscure cases like emoji. 👍

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/strings/split/split_re.hpp:80]: Might be useful to indicate the reverse direction here.  Otherwise the brief reads the same as a function which does the opposite.

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/strings/split/split_re.hpp:149]:
```suggestion
 * s1 = split_record_re(s, "[_ ]")
```

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/strings/split/split_re.hpp:155]:
```suggestion
 * s2 = split_record_re(s, "[ _]", 1)
```

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/strings/split/split_re.hpp:180]: Same comment as above about reverse-ness.

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/strings/split/split_re.hpp:201]:
```suggestion
 * s1 = rsplit_record_re(s, "[_ ]")
```

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (nvdbaranec) [inline cpp/include/cudf/strings/split/split_re.hpp:207]:
```suggestion
 * s2 = rsplit_record_re(s, "[ _]", 1)
```

File: cpp/src/strings/split/split_re.cu
Comment (nvdbaranec) [inline cpp/src/strings/split/split_re.cu:218]: Would it make sense to roll this into `generate_tokens` and have it return the tokens/offsets as a pair?  It already has everything it needs to do the `count_matches` step and the other use case below (`split_record_re`) does the same thing.

Comment (davidwendt) [reply cpp/src/strings/split/split_re.cu:218]: They offsets are not technically generated the same. In the `split_re` the offsets are only temporary working data but in the `split_record_re` they are to be returned. So each is built with a different memory resource as appropriate. Though seemingly minor, I feel this is an important distinction between the two implementations.

**Review by nvdbaranec — APPROVED** [review]

**Comment by davidwendt** [comment]
rerun tests

**Comment by davidwendt** [comment]
This is ready to merge. It has a sufficient number of reviews and all comments have been addressed.
Anyone see a reason to keep the `DO NOT MERGE` label?
Any follow up issues can be addressed in follow up PRs.
There is draft #10185 which is waiting for this to be merged.

**Comment by ttnghia** [comment]
> This is ready to merge. It has a sufficient number of reviews and all comments have been addressed. Anyone see a reason to keep the `DO NOT MERGE` label?

We are testing it with some corner cases. @andygrove are you done with the testing? Can this be merged now?

**Review by andygrove — APPROVED** [review]

**Comment by andygrove** [comment]
> We are testing it with some corner cases. @andygrove are you done with the testing? Can this be merged now?

Yes, this LGTM. I have now approved.

**Review by ttnghia — APPROVED** [review]

File: cpp/include/cudf/strings/split/split_re.hpp
Comment (ttnghia) [inline cpp/include/cudf/strings/split/split_re.hpp:56]: Wait wait. The example may be wrong. The result of each input string should be spreading across row of the output table, not col.

Comment (davidwendt) [reply cpp/include/cudf/strings/split/split_re.hpp:56]: This is consistent with `split` which returns a table. See `split_record_re` perhaps which returns list column.

Comment (ttnghia) [reply cpp/include/cudf/strings/split/split_re.hpp:56]: Humm, so the table above is actually not displayed as a table, instead it is displayed as a "list of columns"  right? Each column is written horizontally? If so, that's confusing as when saying "table" the rows display horizontally instead.

Comment (davidwendt) [reply cpp/include/cudf/strings/split/split_re.hpp:56]: The pattern was to match the example which showed the input `s` column horizontally.

Comment (ttnghia) [reply cpp/include/cudf/strings/split/split_re.hpp:56]: Got it. Thanks.

**Comment by davidwendt** [comment]
@gpucibot merge

**Pull request merged**
